### ISSUE TITLE
ASSERT("string") and method call on null in src/pal/src/init/sxs.cpp

### ISSUE THREAD

**Issue by dkreuter** [dotnet/coreclr#90]
There are multiple occurrences of ASSERT wrongly used as logging function in src/pal/src/init/sxs.cpp. String pointers are always non-null, thus the assertions never fail.

     44: ASSERT("PAL function called on a thread unknown to this PAL\n");
    180: ASSERT("PAL_Reenter called on a thread unknown to this PAL\n");
    207: ASSERT("PAL_Reenter called on a thread unknown to this PAL\n");
    242: ASSERT("PAL_ReenterForEH called on a thread unknown to this PAL\n");

Also, even though there is a null check on pThread in line 178, the following call is executed regardless:

    185: pThread->Enter(boundary);

By the way, it would be nice to have a developer IRC channel.

EDIT: I found a similar pattern in src/pal/src/shmemory/shmemory.cpp

    2554: if ( pshmNew == NULL )
    2556:     ASSERT( "pshmNew should not be NULL\n" );
    ...
    2561: pshmNew->ShmNext = SHMGetInfo( SIID_NAMED_OBJECTS );

Line 2556 is a no-op in debug and release build and the null-check doesn't prevent the access of ShmNext.

**Comment by janvorli** [comment]
I understand it looks strange, but ASSERT in the PAL is not checking any condition, but unconditionally asserting and printing the parameter message. See src\pal\src\include\pal\dbgmsg.h@368.
To assert on a condition, there are _ASSERT, _ASSERTE and _ASSERT_MSG macros.

**Comment by dkreuter** [comment]
Oh, o_O, okay. That's unconventional. This definition is local to the PAL I guess. In other parts of the code-base ASSERT and _ASSERTE are the same:

    src/debug/di/rspriv.h:  #define ASSERT(x) _ASSERTE(x)
    src/debug/ee/debugger.h:#define ASSERT(x) _ASSERTE(x)
    src/gc/sample/gcenv.h:  #define ASSERT(_expr) assert(_expr)
    src/vm/hash.h:          #define ASSERT _ASSERTE
    src/palrt/shlwapip.h:   #define ASSERT          _ASSERTE
    src/gc/gcpriv.h:        #define ASSERT _ASSERTE

What about the null deref situations, are they considered to fatal to handle?

**Comment by janvorli** [comment]
Regarding the sxs.cpp, it is a developer's error to call the function on a wrong thread. As such, it should be caught during debugging and fixed. Or maybe the PAL data structures got into an inconsistent state due to a bug in the PAL. Even if we returned an error status here, there would be no reasonable way for the developer to recover from that - other than reporting an error or asserting. Trying to handle it somehow would lead to a problem few steps further when another call would expect a correct state of the system.
As for the SHMAddNamedObject in the shmemory.cpp, it is the same case. I also don't see a reasonable way to recover. Either the developer has called the method with an invalid parameter or the consistency of the PAL data structures was compromised. If we returned an error status and the developer wanted to handle that - what could he / she do other than fail here?
